File: backend/app/models/publish_task.py
from __future__ import annotations

import enum
import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base
from typing import Optional


class PublishTaskStatus(str, enum.Enum):
    pending = "pending"
    publishing = "publishing"
    published = "published"
    failed = "failed"


class PublishTask(Base):
    __tablename__ = "publish_tasks"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    content_id: Mapped[str] = mapped_column(String(36), ForeignKey("contents.id"), nullable=False)
    account_id: Mapped[str] = mapped_column(String(36), ForeignKey("accounts.id"), nullable=False)
    status: Mapped[PublishTaskStatus] = mapped_column(
        String(20), default=PublishTaskStatus.pending, nullable=False
    )
    scheduled_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    published_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    error_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    retry_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.now, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.now, onupdate=datetime.now, nullable=False
    )

    content = relationship("Content", back_populates="publish_tasks")
    account = relationship("Account", back_populates="publish_tasks")

File: backend-go/models/publish_task.go
package models

import (
	"time"
)

const (
	PublishTaskStatusPending    = "pending"
	PublishTaskStatusPublishing = "publishing"
	PublishTaskStatusPublished  = "published"
	PublishTaskStatusFailed     = "failed"
)

type PublishTask struct {
	ID           string     `orm:"column(id);pk;size(36)" json:"id"`
	ContentID    string     `orm:"column(content_id);size(36);index" json:"content_id"`
	AccountID    string     `orm:"column(account_id);size(36);index" json:"account_id"`
	Status       string     `orm:"column(status);size(20);default(pending)" json:"status"`
	ScheduledAt  *time.Time `orm:"column(scheduled_at);type(datetime);null" json:"scheduled_at"`
	PublishedAt  *time.Time `orm:"column(published_at);type(datetime);null" json:"published_at"`
	ErrorMessage string     `orm:"column(error_message);type(text);null" json:"error_message"`
	RetryCount   int        `orm:"column(retry_count);default(0)" json:"retry_count"`
	CreatedAt    time.Time  `orm:"column(created_at);auto_now_add;type(datetime)" json:"created_at"`
	UpdatedAt    time.Time  `orm:"column(updated_at);auto_now;type(datetime)" json:"updated_at"`
}

func (p *PublishTask) TableName() string {
	return "publish_tasks"
}
